---
layout: post
title: 자바스크립트의 프로미스 기초
---


## 1. 프로미스의 정의
자바스크립트에서 비동기 처리를 해주는 함수

`new Promise()`로 프로미스를 생성하고, 프로미스 내부의 비동기 처리가 끝나면  어떤 일을 실행하게 할 수 있다.

프로미스는 크게 3가지 중요한 내용이 있다. 
1. Producer, Consumer를 구분하는 것 
2. 프로미스의 상태를 아는 것
3. 프로미스 체인



## 2. Producer, Consumer
{% highlight js %}
//1. Producer: 복잡한 일을 수행하는 부분
const promise = new Promise((resolve, reject) => {
    console.log("doing something...");
    setTimeout(() => {
        resolve('BeaChu');
    }, 2000);
});

//2. Consumer: 복잡한 일이 끝나고 실행되는 부분
promise
  .then(value => {
      console.log(value); //실행되는 부분
  })
  .catch(error => {
      console.log(error); //resolve 대신 reject가 있었다면 실행될 부분
  });
{% endhighlight %}
Producer는 프로미스의 객체를 생성하는 부분으로, 복잡한 일을 비동기적으로 처리하는 부분이다. 또한 Consumer는 Producer의 실행을 마치고 나면 실행될 부분이다.

예를 들면 Producer는 서버에서 많은 정보를 받아오는 부분, Consumer는 정보를 받아오고 나서 사용자에게 알려주는 부분이라고 생각할 수 있을 것이다.



## 3. 프로미스의 상태
{% highlight js %}
//1. Pending(대기)
new Promise(function(resolve, reject) {
  // ...
});

//2. Fulfilled(이행, 완료)
function getData() {
  return new Promise(function(resolve, reject) {
   var data = 100;
   resolve(data);
  });
 }
 
 // resolve()의 결과 값 data를 resolvedData로 받음
 getData().then(function(resolvedData) {
  console.log(resolvedData); // 100
 });

//3. Rejected(실패)
function getData() {
  return new Promise(function(resolve, reject) {
   reject(new Error("Request is failed"));
  });
 }

 // reject()의 결과 값 Error를 err에 받음
 getData().then().catch(function(err) {
  console.log(err); // Error: Request is failed
 });
{% endhighlight %}

위 코드에서 볼 수 있는 것처럼 프로미스의 상태는

-  Pending(대기): 비동기 처리가 완료되지 않은 상태
-  Fulfiiled(이행): 비동기 처리가 완료된 상태
-  Rejected(실패): 비동기 처리가 실패하거나 오류가 발생한 상태

이렇게 3가지를 가진다.


## 4. 프로미스 연결하기
프로미스의 특징 중 하나는 프로미스 여러개를 연결하여 사용할 수 있다는 점이다. 
{% highlight js %}
Promise
  .then((message) => { 
    return new Promise((resolve, reject) => {
      resolve(message);
    });
  })
  .then((message2) => { //첫번째 then에서 resolve된 message를 message2로 받음
    console.log(message2);
    return new Promise((resolve, reject) => {
        resolve(message2);
    });
  })
  .then((message3) => {
     console.log(message3);
  })
  .catch((error) => {
      console.error(error);
  });
  {% endhighlight %}
  
  - 위 예시의 첫번째 then에서 return된 프로미스가 message를 resolve한다. 
  - 즉 프로미스의 Producer가 호출되고 Fulfiiled된 상태가 된다. 
  - 그러면 두번째 then이 실행되고 앞에서 resolve된 message를  message2로 전달 받는다. 
  - 마찬가지로 다음 then에서는 message2를 message3로 전달받는다. 
  - 마지막에 있는 .catch를 통해서 오류를 처리해주고 있다.
  